Transports/com.mlapi.contrib.transport.facepunch/Runtime/FacepunchTransport.cs

```diff
@@ -0,0 +1,363 @@
+        }
+
+        public override void DisconnectRemoteClient(ulong clientId)
+        {
+            if (connectedClients.TryGetValue(clientId, out Client user))
+            {
+                user.connection.Close();
+                connectedClients.Remove(clientId);
+
+                if (LogLevel <= LogLevel.Developer)
+                    NetworkLog.LogInfoServer($"[{nameof(FacepunchTransport)}] - Disconnecting remote client with ID {clientId}.");
+            }
+            else if (LogLevel <= LogLevel.Normal)
+                NetworkLog.LogWarningServer($"[{nameof(FacepunchTransport)}] - Failed to disconnect remote client with ID {clientId}, client not connected.");
+        }
+
+        public override unsafe ulong GetCurrentRtt(ulong clientId)
+        {
+            return 69;
+        }
+
+        public override void Init()
+        {
+            connectedClients = new Dictionary<ulong, Client>();
+            channelSendTypes = new Dictionary<NetworkChannel, SendType>();
+            eventQueue = new Queue<ConnectionEvent>();
+
+            foreach (TransportChannel channel in MLAPI_CHANNELS.Concat(channels))
+            {
+                SendType sendType = channel.Delivery switch
+                {
+                    NetworkDelivery.Reliable => SendType.Reliable,
+                    NetworkDelivery.ReliableFragmentedSequenced => SendType.Reliable,
+                    NetworkDelivery.ReliableSequenced => SendType.Reliable,
+                    NetworkDelivery.Unreliable => SendType.Unreliable,
+                    NetworkDelivery.UnreliableSequenced => SendType.Unreliable,
+                    _ => SendType.Reliable
+                };
+
+                channelSendTypes.Add(channel.Channel, sendType);
+            }
+        }
+
+        public override void Shutdown()
+        {
+            try
+            {
+                if (LogLevel <= LogLevel.Developer)
+                    NetworkLog.LogInfoServer($"[{nameof(FacepunchTransport)}] - Shutting down.");
+
+                connectionManager?.Close();
+                socketManager?.Close();
+            }
+            catch (Exception e)
+            {
+                if (LogLevel <= LogLevel.Error)
+                    NetworkLog.LogErrorServer($"[{nameof(FacepunchTransport)}] - Caught an exception while shutting down: {e}");
+            }
+        }
+
+        public override unsafe void Send(ulong clientId, ArraySegment<byte> data, NetworkChannel networkChannel)
+        {
+            if (!channelSendTypes.TryGetValue(networkChannel, out SendType sendType))
+                if (!channelSendTypes.TryGetValue(NetworkChannel.DefaultMessage, out sendType))
+                    sendType = SendType.Reliable;
+
+            if (clientId == ServerClientId)
+                fixed (byte* pointer = data.Array, buffer = new byte[data.Count + 1])
+                {
+                    Buffer.MemoryCopy(pointer + data.Offset, buffer, data.Count, data.Count);
```

Comment:
Why do you copy the data to another buffer here instead of passing the pointer directly?